### PR TITLE
Gradient accumulation

### PR DESCRIPTION
**Pull request by LysandreJik**
Congratulations! You've made it this far! Once merged, the article will appear at https://huggingface.co/blog. Official articles 
require additional reviews. Alternatively, you can write a community article following the process [here](https://huggingface.co/blog-explorers).

## Preparing the Article

You're not quite done yet, though. Please make sure to follow this process (as documented [here](https://github.com/huggingface/blog/tree/main?tab=readme-ov-file#how-to-write-an-article-)):

- [x] Add an entry to [_blog.yml](https://github.com/huggingface/blog/blob/main/_blog.yml).
- [x] Add a thumbnail. There are no requirements here, but there is a [template](https://github.com/huggingface/blog?tab=readme-ov-file#how-to-get-a-nice-responsive-thumbnail) if it's helpful.
- [x] Check you use a short title and blog path.
- [ ] Upload any additional assets (such as images) to the Documentation Images [repo](https://huggingface.co/datasets/huggingface/documentation-images/tree/main/blog). This is to reduce bloat in the GitHub base repo when cloning and pulling. Try to have small images to avoid a slow or expensive user experience.
- [x] Add metadata (such as authors) to your `md` file. You can also specify `guest` or `org` for the authors.
- [x] Ensure the publication date is correct.
- [ ] Preview the content. A quick way is to paste the markdown content in https://huggingface.co/new-blog. **Do not click publish**, this is just a way to do an early check.

Here is an example of a complete PR: https://github.com/huggingface/blog/pull/2382

## Getting a Review

Please make sure to get a review from someone on your team or a co-author.
Once this is done and once all the steps above are completed, you should be able to merge.
There is no need for additional reviews if you and your co-authors are happy and meet all of the above.

Feel free to add @osanseviero or @pcuenca as reviewers if you want a final check. They'll be biased toward light reviews
(e.g., check for proper metadata) rather than content reviews unless explicitly asked.

### CODE REVIEW

**Comment by LysandreJik** [comment]
cc @ArthurZucker @muellerzr

File: gradient_accumulation.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+thumbnail: /blog/assets/159_autogptq_transformers/thumbnail.jpg
+authors:
+- user: lysandre
+- user: ArthurZ
+- user: muellerzr
+---
+
+# Fixing Gradient Accumulation
+
+Our friends at Unsloth shared an issue regarding gradient accumulation yesterday that is affecting the transformers Trainer.
```
Comment (muellerzr) [inline gradient_accumulation.md:13]: Should we link to it?

```suggestion
Our friends at Unsloth [shared an issue](https://unsloth.ai/blog/gradient) regarding gradient accumulation yesterday that is affecting the transformers Trainer.
```

File: gradient_accumulation.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+Gradient accumulation is supposed to be mathematically equivalent to full batch training; however, losses did not match between training runs where the setting was toggled on and off.
```
Comment (muellerzr) [inline gradient_accumulation.md:15]:
```suggestion
Gradient accumulation is *supposed* to be mathematically equivalent to full batch training; however, losses did not match between training runs where the setting was toggled on and off.
```

(For emphasis)

File: gradient_accumulation.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+### Where does it stem from?
+
+First of all, inside the modeling code of each model, transformers offers a "default" loss function that's the most typically used loss function by the model architecture. It depends on what the class itself should be used for: question answering, token classification, causal LM, masked LM.
```
Comment (muellerzr) [inline gradient_accumulation.md:19]: Not sure we need the "First of all" here + a few other suggestions

```suggestion
Inside the modeling code of each model, `transformers` offers a "default" loss function that's the most typically used loss function by the model architecture. It is determined by what the class itself should be used for: question answering, token classification, causal LM, masked LM.
```

File: gradient_accumulation.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+This is the default method which is not meant to be customizable: it is only computed when labels, as well as input_ids, are passed sa inputs to the model. The default loss is useful but is limited by design: for anything different being done, then we expect the labels to not be passed directly, and for users to get the logits back from the model and use them to compute the loss outside of the model.
+
```
Comment (muellerzr) [inline gradient_accumulation.md:22]:
```suggestion
This is the default method which is not meant to be customizable: it is only computed when `labels` and `input_ids` are passed as inputs to the model. The default loss is useful but is limited **by design**: for anything different being done, then we expect the labels to **not be passed directly, and for users to get the logits back from the model and use them to compute the loss outside of the model.**

```

More emphasis + nit

File: gradient_accumulation.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+However, the transformers Trainer, as well as many others, is heavily leveraging these methods: by the simplicity it offers, it is a double-edged sword. Providing a simple API that becomes different as the use-case differs is not a well-thought API, and we've been caught by surprise ourselves.
```
Comment (muellerzr) [inline gradient_accumulation.md:23]:
```suggestion
However, the transformers Trainer, as well as many Trainers, is heavily leveraging these methods: by the simplicity it offers, it is a double-edged sword. Providing a simple API that becomes different as the use-case differs is not a well-thought API, and we've been caught by surprise ourselves.
```

File: gradient_accumulation.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+### How we're fixing it
+
+To address this issue, we’re changing the way our models and training works in two ways:
+If users are using the “default” loss functions by each model architecture, we will automatically take into account the needed changes when using gradient accumulation on them to make sure the proper loss is reported and utilized, fixing the core issue at hand. 
+To ensure that any future issues with calculating losses might come up in the future won’t block users, we’ll be exposing an API to let users pass in their own loss functions to the `Trainer` directly so they can use their own fix in easily until we have fixed any issues internally and pushed onto a new release.
```
Comment (muellerzr) [inline gradient_accumulation.md:29]:
```suggestion
To address this issue, we’re changing the way our models and training works in two ways:

* If users are using the “default” loss functions by each model architecture, we will automatically take into account the needed changes when using gradient accumulation on them to make sure the proper loss is reported and utilized, fixing the core issue at hand. 
* To ensure that any future issues with calculating losses might come up in the future won’t block users, we’ll be exposing an API to let users pass in their own loss functions to the `Trainer` directly so they can use their own fix in easily until we have fixed any issues internally and pushed onto a new release. 
```

Formatting nit

**Review by muellerzr — APPROVED** [review]
Thanks! LG2M just a few suggestions + nits

File: gradient_accumulation.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+This is the default method which is not meant to be customizable: it is only computed when `labels` and `input_ids` are passed as inputs to the model. The default loss is useful but is limited **by design**: for anything different being done, then we expect the labels to **not be passed directly, and for users to get the logits back from the model and use them to compute the loss outside of the model.**
+
+However, the transformers Trainer, as well as many Trainers, is heavily leveraging these methods: by the simplicity it offers, it is a double-edged sword. Providing a simple API that becomes different as the use-case differs is not a well-thought API, and we've been caught by surprise ourselves.
+
+To be precise, for gradient accumulation across token-level tasks like causal LM training, the correct loss should be computed by the total loss across all batches in a gradient accumulation step divided by the total number of all non padding tokens in those batches. This is not the same as the average of the per-batch loss values
```
Comment (ydshieh2) [inline gradient_accumulation.md:26]:
```suggestion
To be precise, for gradient accumulation across token-level tasks like causal LM training, the correct loss should be computed by the total loss across all batches in a gradient accumulation step divided by the total number of all non padding tokens in those batches. This is not the same as the average of the per-batch loss values.
```

Comment (ydshieh2) [reply gradient_accumulation.md:26]: it's @ydshieh don't worry

File: gradient_accumulation.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
+
+---
+title: "Fixing Gradient Accumulation" 
+thumbnail: /blog/assets/159_autogptq_transformers/thumbnail.jpg
```
Comment (pcuenca) [inline gradient_accumulation.md:4]:
```suggestion
thumbnail: /blog/assets/gradient_accumulation/gradient_accumulation.png
```

File: gradient_accumulation.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+- user: lysandre
+- user: ArthurZ
+- user: muellerzr
+- user: ydshieh
+- user: BenjaminB
+---
+
+# Fixing Gradient Accumulation
+
+Our friends at Unsloth [shared an issue](https://unsloth.ai/blog/gradient) regarding gradient accumulation yesterday that is affecting the transformers Trainer.
```
Comment (ArthurZucker) [inline gradient_accumulation.md:15]:
```suggestion
Our friends at Unsloth [shared an issue](https://unsloth.ai/blog/gradient) regarding gradient accumulation yesterday that is affecting the transformers Trainer. The initial report comes from @bnjmn_marie (kudos to him!).
```

File: gradient_accumulation.md (hunk 8 above)
Comment (LysandreJik) [reply gradient_accumulation.md:4]: damn I was sure I had changed that 🤦 I'm a noob, thanks Pedro

File: gradient_accumulation.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+### Where does it stem from?
+
+Inside the modeling code of each model, `transformers` offers a "default" loss function that's the most typically used loss function by the model architecture. It is determined by what the class itself should be used for: question answering, token classification, causal LM, masked LM.
+
+This is the default method which is not meant to be customizable: it is only computed when `labels` and `input_ids` are passed as inputs to the model. The default loss is useful but is limited **by design**: for anything different being done, we expect the labels to **not be passed directly, and for users to get the logits back from the model and use them to compute the loss outside of the model.**
+
+However, the transformers Trainer, as well as many Trainers, is heavily leveraging these methods: by the simplicity it offers, it is a double-edged sword. Providing a simple API that becomes different as the use-case differs is not a well-thought out API, and we've been caught by surprise ourselves.
```
Comment (ArthurZucker) [inline gradient_accumulation.md:25]:
```suggestion
However, the transformers Trainer, as well as many Trainers, heavily leverage these methods because of the simplicity it offers: it is a double-edged sword. Providing a simple API that becomes different as the use-case differs is not a well-thought out API, and we've been caught by surprise ourselves.
```

File: gradient_accumulation.md [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+To be precise, for gradient accumulation across token-level tasks like causal LM training, the correct loss should be computed by the total loss across all batches in a gradient accumulation step divided by the total number of all non padding tokens in those batches. This is not the same as the average of the per-batch loss values.
+
```
Comment (ArthurZucker) [inline gradient_accumulation.md:28]:
```suggestion
The fix is quite simple, see the following:
```diff
def ForCausalLMLoss(logits, labels, vocab_size, **kwargs):
    # Upcast to float if we need to compute the loss to avoid potential precision issues
    logits = logits.float()
    # Shift so that tokens < n predict n
    shift_logits = logits[..., :-1, :].contiguous()
    shift_labels = labels[..., 1:].contiguous()

    # Flatten the tokens
    shift_logits = shift_logits.view(-1, vocab_size)
    shift_labels = shift_labels.view(-1)
    # Enable model parallelism
    shift_labels = shift_labels.to(shift_logits.device)

    num_items = kwargs.pop("num_items", None)
+        loss = nn.functional.cross_entropy(shift_logits, shift_labels, ignore_index=-100, reduction="sum")
+        loss = loss / num_items
-        loss = nn.functional.cross_entropy(shift_logits, shift_labels, ignore_index=-100)
    return loss
```
```

File: gradient_accumulation.md [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+### How we're fixing it
+
+To address this issue, we’re changing the way our models and training works in two ways:
+
+* If users are using the “default” loss functions by each model architecture, we will automatically take into account the needed changes when using gradient accumulation on them to make sure the proper loss is reported and utilized, fixing the core issue at hand. 
+* To ensure that any future issues with calculating losses won’t block users, we’ll be exposing an API to let users pass in their own loss functions to the `Trainer` directly so they can use their own fix easily until we have fixed any issues internally and made a new transformers release. 
+
+### When will it land
```
Comment (ArthurZucker) [inline gradient_accumulation.md:36]: All model that inherit from `PreTrainedModel` now have a `loss_function` property, which is determined by either: 
- the `config.loss_type`: this is to make sure anyone can use his custom loss. You can do this by modifying the `LOSS_MAPPING`:
```python
def my_super_loss(logits, labels):
    return loss = nn.functional.cross_entropy(logits, labels, ignore_index=-100)


LOSS_MAPPING["my_loss_type"] = my_super_loss
```
- the name of the class: most of our classes have names related to tasks, like `ForObjectDetection`, or `ForMaskedLM`.

**Review by ArthurZucker — APPROVED** [review]

File: gradient_accumulation.md [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+To be precise, for gradient accumulation across token-level tasks like causal LM training, the correct loss should be computed by the total loss across all batches in a gradient accumulation step divided by the total number of all non padding tokens in those batches. This is not the same as the average of the per-batch loss values.
+
+### How we're fixing it
+
+To address this issue, we’re changing the way our models and training works in two ways:
+
+* If users are using the “default” loss functions by each model architecture, we will automatically take into account the needed changes when using gradient accumulation on them to make sure the proper loss is reported and utilized, fixing the core issue at hand. 
+* To ensure that any future issues with calculating losses won’t block users, we’ll be exposing an API to let users pass in their own loss functions to the `Trainer` directly so they can use their own fix easily until we have fixed any issues internally and made a new transformers release.
```
Comment (pcuenca) [inline gradient_accumulation.md:53]: Are there any references in the documentation about using your own loss function?

**Review by pcuenca — APPROVED** [review]
Awesome, added some minor nits.

File: gradient_accumulation.md (hunk 13 above)
Comment (LysandreJik) [reply gradient_accumulation.md]: Not yet I believe, but we'll add some! Thanks!

**Comment by LysandreJik** [comment]
Thanks all, merging!

File: gradient_accumulation.md (hunk 13 above)
Comment (pcuenca) [reply gradient_accumulation.md]: 👍 I also meant for the classic way to compute the loss externally, perhaps we can link to it when we discuss that developers were expected to do that, and then maybe add an action item here to cover all the options in the documentation.
